pingcap/tidb
# type/compatibility: fix show create table output of generated column

### What problem does this PR solve? 
fix incompatible output of `show create table` for generated column

### What is changed and how it works?
Change method for restoring `ast.ColumnOptionGenerated` from `Format` to `Restore`
Change related `cast` expression in parser
Related issue #9457

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
```sql
mysql> 
mysql> CREATE TABLE t20( first_name varchar(10), last_name varchar(10), full_name varchar(255) AS (CONCAT(first_name,' ' ,last_name) ));
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t20;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                      |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t20   | CREATE TABLE `t20` (
  `first_name` varchar(10) DEFAULT NULL,
  `last_name` varchar(10) DEFAULT NULL,
  `full_name` varchar(255) GENERATED ALWAYS AS (concat(`first_name`, ' ', `last_name`)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                                                |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta-170-ga64f3092f-dirty
Git Commit Hash: a64f3092f49fff6ab0538dc3bdaf43cb53553a3f
Git Branch: show_create_table_virtual
UTC Build Time: 2019-03-07 04:05:05
GoVersion: go version go1.11.2 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 
```

Code changes

 - Has exported function/method change
No.
 - Has exported variable/fields change
No.
 - Has interface methods change
No.
 - Has persistent data change
No.

Side effects

 - Possible performance regression
No.
 - Increased code complexity
No.
 - Breaking backward compatibility
No.

Related changes

 - Need to cherry-pick to the release branch
No.
 - Need to update the documentation
No.
 - Need to update the `tidb-ansible` repository
No.
 - Need to be included in the release note
No.
